### PR TITLE
 scripts/release-notes: enable amending past release notes

### LINKED ISSUE #42163: scripts/release-notes: make it possible to edit a release note in a different commit
Following up to a discussion with @BramGruneir 

This issue is to propose an extension to the release note extraction script.

The extension would make it possible to amend a release note from the past using a new commit message. The script would be changed to recognize the amend in the later commit and change the original release note in the notes report that it produces.

For example, say some engineer A wrote the following release note in April on PR 12345:

```
Release note (bug fix): Added a column to crdb_internal.wabble. 
I think it was broken. Not sure.
```

Then in May someone notices the release note was incomplete. This person can then use:

```
$ git commit --allow-empty
```

then fill in the following release note:

```
Release note (edit #12345, bug fix): The new column `world` was added to
`crdb_internal.wabble`. This makes it possible to join this table with
`crdb_internal.wabble`, as was originally intended. This limitation was first
introduced in version 31.2.
```

This then causes the release note script to produce the following output:

```
Section "Bug fixes":

- Added a column to crdb_internal.wabble. I think it was broken. Not sure.
  AMENDED:
  The new column `world` was added to `crdb_internal.wabble`. This makes it
  possible to join this table with `crdb_internal.wabble`, as was originally
  intended. This limitation was first introduced in version 31.2. [#12345]
```

### PR DESCRIPTION
**Pull request by knz**
Fixes #42163.
First two commits from #42339.

Previously, the only moment when a team member could contribute a
release note to a change was on the PR implementing the change.

This is because the script to extract release notes maps
release notes to the PRs enclosing the note's own commit.

The present change extends the mechanism by enabling authors to revise
release notes *at an arbitrary later point in the repository*,
including even when the revision is out of the selection range for the
release note script.

For example:

   ```
   Revised release note [#123] (bug fix): blah blah...
   ```

or:

   ```
   Revised release note [abd1123f] (bug fix): blah blah...
   ```

This instructs the script to add a "NOTE REVISION" entry to the `bug
fix` release note present in PR #123 or commit abd1123f.

Example use:

```shell
$ python3 scripts/release-notes.py \
  --from  c107719 \
  --until 28b447a
```

(This selects automatically `--revisions-until=master`, but this can
be overridden if needed.)

Example output, using the release note revision on this
commit below:

```
- Revert ability to create second-precision
  (0-precision) timestamp datatypes, which inadvertently caused a backward
  compatibility issue that automatically made previously-existing
  timestamp columns from 19.1 into second-precision columns.
  - NOTE REVISION [424a225c7][424a225c7]
    The newly-introduced
    support for second-precision columns (e.g. `TIMESTAMP(0)`) was
    removed, due to its unintended backward-incompatible applicaton
    to all previously-defined columns without a precision. [#42285][#42285] [5a4e077][5a4e077]
```

Moreover, if the target PR or commit SHA did not have a release note
in the same category already, the note revision is inserted as a fresh
release note.

Revised release note [#42285] (bug fix): The newly-introduced
support for second-precision columns (e.g. `TIMESTAMP(0)`) was
removed, due to its unintended backward-incompatible applicaton
to all previously-defined columns without a precision.

Release note: None

### CODE REVIEW

**Comment by knz** [comment]
cc @BramGruneir this is what we discussed. With this change, it becomes possible to document known limitations, backward-incompatible changes, etc, "after the fact".

**Comment by knz** [comment]
@chrisseto if you could review the code change that would be useful.

**Review by jseldess — APPROVED** [review]
LGTM in general, though someone else should review the code. One question, though: If someone revises a release note text after the release goes out including the original text, how are we on Docs supposed to know about that and act on the changed text?

**Comment by knz** [comment]
> If someone revises a release note text after the release goes out including the original text, how are we on Docs supposed to know about that and act on the changed text?

I hadn't thought about that. Maybe the script can emit a separate section with "note revisions in the current commit selection that target notes out of this range"?

**Comment by jseldess** [comment]
@knz, I don't understand what you're suggesting. Do you think devs will generally do this soon after the commit is merged, or weeks after? If the former, I think we'll be ok. If the latter, we need to establish a process by which Docs finds out about this, and I'm not sure what that process should be.

**Review by chrisseto — APPROVED** [review]
> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42344)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell and @chrisseto)

**Comment by knz** [comment]
> @knz, I don't understand what you're suggesting. Do you think devs will generally do this soon after the commit is merged, or weeks after? 

I think both can happen.

>  we need to establish a process by which Docs finds out about this, and I'm not sure what that process should be.

So it's possible for the script to pop up a section in the report that indicates that there were revisions that have not been processed yet. This report would be generated when producing release notes for a new release but pertain to the release before that. Would that work?

**Comment by jseldess** [comment]
I like that idea, @knz. We will probably need to make it a flag to explicitly point to the prior release since we alternate between 19.2.x, 20.1.x, etc.

**Comment by knz** [comment]
> So it's possible for the script to pop up a section in the report that indicates that there were revisions that have not been processed yet. This report would be generated when producing release notes for a new release but pertain to the release before that. Would that work?

I tried to do this and it turns out this is not really possible.

In fact the approach proposed in the script currently is simply inadequate for release branches: the PR numbers that are targeted by note revisions are those from the original PRs, not the backport PR numbers which are listed in the git log. Today the two are not linked and so the script is unable to attach revisions to the release branch properly.

I'm not sure yet how to proceed, will need to think about this a bit more.

**Comment by knz** [comment]
I wasn't able to find a solution to the problem identified above.
We probably need to address this problem in a different way.

**Pull request closed without merging**
